ocaml/gencert/pem.ml

```diff
@@ -60,24 +62,35 @@ let key =
   key_footer kind *> return (String.concat "" [header kind; body; footer kind])
   <?> "key"
 
-let line = take_till is_eol *> end_of_line
-
-(* try to read a key, or skip a line and try again *)
-let until_key = fix (fun m -> key <|> line *> m) <?> "until_key"
-
 let cert =
   cert_header >>= fun hd ->
   data >>= fun body ->
   cert_footer >>= fun tl -> return (String.concat "" [hd; body; tl]) <?> "cert"
 
-(* try to read a cert, or skip a line and try again *)
-let until_cert = fix (fun m -> cert <|> line *> m) <?> "until_cert"
+let line = take_till is_eol *> end_of_line
+
+let any_block =
+  cert >>= (fun c -> return (Cert c)) <|> (key >>= fun k -> return (Key k))
+
+(* this skips over junk until we succeed finding the next block *)
+let block = fix (fun m -> any_block <|> line *> m) <?> "until_block"
```

Comment:
Maybe not a concern here, since the files are usually small, but should we use the 'commit' function from Angstrom to prevent backtracking too much? 
Otherwise with N lines it will attempt 2^N ways to parse it.

Angstrom is not like a usual, efficient, parser that construct a table to scan the input in ~linear time. Worst case it can take 2^N time, where N is the number of alternatives.